File: src/main/java/com/vimeo/networking/model/FeedItem.java
package com.vimeo.networking.model;

import java.io.Serializable;
import java.util.Date;

/**
 * Created by zetterstromk on 6/24/15.
 */
public class FeedItem implements Serializable {

    private static final long serialVersionUID = -8744477085158366576L;

    //TODO I have only seen like/channel types, we need more types [KZ]
    public enum AttributionType {
        UPLOAD,
        LIKE,
        CHANNEL,
        GROUP,
        TAG,
        CREDIT,
        SHARE,
        NONE
    }

    public String uri;
    public Video clip;
    public String type;
    public Date time;
    public User user;   // from like type
    public Channel channel; // from channel type
    public Tag tag;
    public Group group;

    public AttributionType getType() {
        if (type.equalsIgnoreCase("channel")) {
            return AttributionType.CHANNEL;
        } else if (type.equalsIgnoreCase("like")) {
            return AttributionType.LIKE;
        } else if (type.equalsIgnoreCase("upload")) {
            return AttributionType.UPLOAD;
        } else if (type.equalsIgnoreCase("tag")) {
            return AttributionType.TAG;
        } else if (type.equalsIgnoreCase("group")) {
            return AttributionType.GROUP;
        } else if (type.equalsIgnoreCase("appearance")) {
            return AttributionType.CREDIT;
        } else if (type.equalsIgnoreCase("share")) {
            return AttributionType.SHARE;
        }

        return AttributionType.NONE;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        FeedItem that = (FeedItem) o;

        return ((this.clip.uri != null && that.clip.uri != null) && this.clip.uri.equals(that.clip.uri));
    }

    @Override
    public int hashCode() {
        return this.clip.uri != null ? this.clip.uri.hashCode() : 0;
    }

}

File: models/src/main/java/com/vimeo/networking2/FeedItem.kt
@file:JvmName("FeedItemUtils")

package com.vimeo.networking2

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import com.vimeo.networking2.enums.AttributionType
import com.vimeo.networking2.enums.asEnum
import java.util.Date

/**
 * An item in the user's feed.
 *
 * @param category The category that this event occurred for. This will be present for only [AttributionType.CATEGORY]
 * feed item type.
 * @param channel The channel that this event occurred for. This will be present for only [AttributionType.CHANNEL] feed
 * item type.
 * @param group The group that this event occurred for. This will be present for only [AttributionType.GROUP] feed item
 * type.
 * @param metadata The feed item's metadata.
 * @param rawType Feed item type. See [FeedItem.type].
 * @param tag The tag that this event occurred for. This will be present for only [AttributionType.TAG] feed item type.
 * @param time Time that the event occurred.
 * @param video Video associated with this feed item.
 * @param uri The uri for the [FeedItem].
 * @param user The user that this event occurred for. This will be present for [AttributionType.LIKE],
 * [AttributionType.APPEARANCE], and [AttributionType.SHARE] activity types.
 */
@JsonClass(generateAdapter = true)
data class FeedItem(

    @Json(name = "category")
    val category: Category? = null,

    @Json(name = "channel")
    val channel: Channel? = null,

    @Json(name = "group")
    val group: Group? = null,

    @Json(name = "metadata")
    val metadata: MetadataConnections<FeedItemConnections>? = null,

    @Json(name = "type")
    val rawType: String? = null,

    @Json(name = "tag")
    val tag: Tag? = null,

    @Json(name = "time")
    val time: Date? = null,

    @Json(name = "clip")
    val video: Video? = null,

    @Json(name = "uri")
    val uri: String? = null,

    @Json(name = "user")
    val user: User? = null

)

/**
 * @see FeedItem.rawType
 * @see AttributionType
 */
val FeedItem.type: AttributionType
    get() = rawType.asEnum(AttributionType.UNKNOWN)
